### PR TITLE
feat(chat): Chatbar markdown preview

### LINKED ISSUE #1367: feat(chat): allow realtime chat preview in text input
It would be nice to have realtime markdown preview in the chat input so the user does not need to send a message to see how it looks like

### LINKED ISSUE #1737: bug(chat): appears 0/1024 when copy pasting code
STR:
- go to chat
- copy paste the following

```python
def hello_world():
    print("Hello, World!")
```
appears 0/1024

Captura de ecrã 2024-01-19, às 14 39 28

macOS, m1

### PR DESCRIPTION
**Pull request by Flemmli97**
### What this PR does 📖

- Adds markdown preview to the chatbar and editing textarea
- Uses this js library which is a wrapper around the codemirror js lib:
https://github.com/Flemmli97/markdown-editor

### Which issue(s) this PR fixes 🔨

- Resolve #1367
- Resolve #1737

### CODE REVIEW

**Comment by phillsatellite** [comment]
@Flemmli97 Hey Kevin! I had some issues trying to get the markdown to appear in the chatbar on any of the OS's 

https://github.com/Satellite-im/Uplink/assets/93608357/16afc3a0-f175-4d3e-b803-136de570262e

**Comment by Flemmli97** [comment]
fixed

**Review by stavares843 — CHANGES_REQUESTED** [review]
in dev sending this works

```## Escaping Characters

To display a literal asterisk \* or underscore \_ in Markdown, use a backslash \ before the character.
``` 

Captura de ecrã 2024-01-17, às 22 43 44


but on this branch does not work and returns the following error on the log

Captura de ecrã 2024-01-17, às 22 44 41


i can copy paste but i cant sent it and also says 0/1024 on the chars, and the send button does not work

Captura de ecrã 2024-01-17, às 22 41 52

**Comment by luisecm** [comment]
Looking at the automated test results:

1. The following test "Favorites - Open Chat conversations with multiple users on Sidebar - Chat User B" is failing because test runner is failing due to timeout after retrying to type on Chat Screen - Text input field. Test runner is typing the text "Testing" but the value assigned to input field is "Testing " (one extra space added).  Is this extra space something we should keep?
image

I am not sure if it is related to the following line 419 on kit/src/elements/textarea/mod.rs] file:
image

2. The following test "Input Bar - Chars Counter on Input Bar displays 0/1024 before typing a text" is failing because the UI locator for inputCharMaxText is being modified in the DOM tree (I see a new text element "Selection deleted", which is moving the existing UI locators used on tests). This is something I can quickly fix when this PR is merged.
image

**Comment by stavares843** [comment]
thanks @luisecm 🔨

**Comment by stavares843** [comment]
btw @Flemmli97 another char where it happens the previous comment as well


```
$ \scriptsize{\text{alert('XSS Attack')}} $
```

Captura de ecrã 2024-01-17, às 23 15 14

**Comment by Flemmli97** [comment]
> Looking at the automated test results:
> 
> 1. The following test "Favorites - Open Chat conversations with multiple users on Sidebar - Chat User B" is failing because test runner is failing due to timeout after retrying to type on Chat Screen - Text input field. Test runner is typing the text "Testing" but the value assigned to input field is "Testing " (one extra space added).  Is this extra space something we should keep?
>    image
> 
> I am not sure if it is related to the following line 419 on kit/src/elements/textarea/mod.rs] file: image
> 
> 2. The following test "Input Bar - Chars Counter on Input Bar displays 0/1024 before typing a text" is failing because the UI locator for inputCharMaxText is being modified in the DOM tree (I see a new text element "Selection deleted", which is moving the existing UI locators used on tests). This is something I can quickly fix when this PR is merged.
>    image

hmm idk why there is a whitespace. testing it in the app and looking at the dom there is no whitespace added there

**Comment by luisecm** [comment]
> > Looking at the automated test results:
> > 
> > 1. The following test "Favorites - Open Chat conversations with multiple users on Sidebar - Chat User B" is failing because test runner is failing due to timeout after retrying to type on Chat Screen - Text input field. Test runner is typing the text "Testing" but the value assigned to input field is "Testing " (one extra space added).  Is this extra space something we should keep?
> >    image
> > 
> > I am not sure if it is related to the following line 419 on kit/src/elements/textarea/mod.rs] file: image
> > 
> > 2. The following test "Input Bar - Chars Counter on Input Bar displays 0/1024 before typing a text" is failing because the UI locator for inputCharMaxText is being modified in the DOM tree (I see a new text element "Selection deleted", which is moving the existing UI locators used on tests). This is something I can quickly fix when this PR is merged.
> >    image
> 
> hmm idk why there is a whitespace. testing it in the app and looking at the dom there is no whitespace added there

I already added one PR on the testing repo that addresses the changes required, you can see the status of tests execution on here: https://github.com/Satellite-im/testing-uplink/pull/606

**Comment by Flemmli97** [comment]
> > > Looking at the automated test results:
> > > 
> > > 1. The following test "Favorites - Open Chat conversations with multiple users on Sidebar - Chat User B" is failing because test runner is failing due to timeout after retrying to type on Chat Screen - Text input field. Test runner is typing the text "Testing" but the value assigned to input field is "Testing " (one extra space added).  Is this extra space something we should keep?
> > >    image
> > > 
> > > I am not sure if it is related to the following line 419 on kit/src/elements/textarea/mod.rs] file: image
> > > 
> > > 2. The following test "Input Bar - Chars Counter on Input Bar displays 0/1024 before typing a text" is failing because the UI locator for inputCharMaxText is being modified in the DOM tree (I see a new text element "Selection deleted", which is moving the existing UI locators used on tests). This is something I can quickly fix when this PR is merged.
> > >    image
> > 
> > 
> > hmm idk why there is a whitespace. testing it in the app and looking at the dom there is no whitespace added there
> 
> I already added one PR on the testing repo that addresses the changes required, you can see the status of tests execution on here: [Satellite-im/testing-uplink#606](https://github.com/Satellite-im/testing-uplink/pull/606)

does that address both points or only the 2.?

**Comment by luisecm** [comment]
> > > > Looking at the automated test results:
> > > > 
> > > > 1. The following test "Favorites - Open Chat conversations with multiple users on Sidebar - Chat User B" is failing because test runner is failing due to timeout after retrying to type on Chat Screen - Text input field. Test runner is typing the text "Testing" but the value assigned to input field is "Testing " (one extra space added).  Is this extra space something we should keep?
> > > >    image
> > > > 
> > > > I am not sure if it is related to the following line 419 on kit/src/elements/textarea/mod.rs] file: image
> > > > 
> > > > 2. The following test "Input Bar - Chars Counter on Input Bar displays 0/1024 before typing a text" is failing because the UI locator for inputCharMaxText is being modified in the DOM tree (I see a new text element "Selection deleted", which is moving the existing UI locators used on tests). This is something I can quickly fix when this PR is merged.
> > > >    image
> > > 
> > > 
> > > hmm idk why there is a whitespace. testing it in the app and looking at the dom there is no whitespace added there
> > 
> > 
> > I already added one PR on the testing repo that addresses the changes required, you can see the status of tests execution on here: [Satellite-im/testing-uplink#606](https://github.com/Satellite-im/testing-uplink/pull/606)
> 
> does that address both points or only the 2.?

It's addressing both points. Since I did not find any visible issue while testing your PR and point #1 could be something Dioxus related, I went ahead and added an update to the tests that will take care of it

**Comment by stavares843** [comment]
when sending this

```
> This is a blockquote.
> It can span multiple lines.
```


it sends fine

Captura de ecrã 2024-01-19, às 14 47 30


but the preview appears like this

Captura de ecrã 2024-01-19, às 14 47 41


the first line is the same as the second in terms of content but layout wise is different


macOS, m1

**Comment by Flemmli97** [comment]
fixed

**Review by stavares843 — APPROVED** [review]

**Comment by Flemmli97** [comment]
before merging will look into #1737 first

**Comment by stavares843** [comment]
sounds good 🔨

**Comment by Flemmli97** [comment]
should be fixed now

**Comment by stavares843** [comment]
thanks 🎉

File: kit/src/elements/textarea/mod.rs [hunk 1]
```diff
@@ -279,3 +294,200 @@ pub fn Input<'a>(cx: Scope<'a, Props<'a>>) -> Element<'a> {
...
+                        if let Some(e) = onkeyup {
+                            e.call(evt.code());
+                        }
+                    },
+                    onkeydown: move |evt| {
+                        // Note for some reason arrow key events are not forwarded to here
+                        // HACK(Linux): Allow copy and paste files for Linux 
+                        if cfg!(target_os = "linux") && evt.code() == Code::KeyV && evt.modifiers() == Modifiers::CONTROL {
+                            if let Some(e) = on_paste_keydown {
+                                log::debug!("paste");
```
Comment (stavares843) [inline kit/src/elements/textarea/mod.rs]: do we need the log?

File: kit/src/elements/textarea/mod.rs [hunk 2]
```diff
@@ -279,3 +294,200 @@ pub fn Input<'a>(cx: Scope<'a, Props<'a>>) -> Element<'a> {
...
+        to_owned![listener_data, eval, value];
+        let rich_editor: String = include_str!("./rich_editor_handler.js")
+            .replace("$EDITOR_ID", &id2)
+            .replace("$AUTOFOCUS", &(!cx.props.ignore_focus).to_string())
+            .replace("$INIT", &value.replace('"', "\\\"").replace('\n', "\\n"));
+        async move {
+            if let Ok(eval) = eval(&rich_editor) {
+                loop {
+                    if let Ok(val) = eval.recv().await {
+                        log::debug!("val {val}");
```
Comment (stavares843) [inline kit/src/elements/textarea/mod.rs]: do we need the log?

**Review by dariusc93 — APPROVED** [review]

**Pull request merged**